### DISCUSSION TITLE
How to take off manually in a real-time simulation?

### DISCUSSION THREAD

**Discussion by hesetone** [JSBSim-Team/jsbsim#768] [Q&A]
Hi:
Thanks for your great work in JSBSim. I am trying to get a real-time simulation. In my project, I setup a workbench program named `wbControl` by Qt, which could send commands to `JSBSim.exe` in the form of `set fcs/elevator-cmd-norm 1`,`set propulsion/engine[0]/set-running 1`,`set fcs/throttle-cmd-norm` etc. My purpose is sending commands to control a take off activity, I want to control its take-off just by commands when the plane is on the ground and at rest. 

1, I choose [jsbsim/aircraft/737/737.xml](https://github.com/JSBSim-Team/jsbsim/tree/master/aircraft/737) as the test aircraft, [jsbsim/aircraft/737/reset00.xml](url) as my initial files.

2, I write commands in the order below, these will be send to `JSBSim.exe` via socket(IP:5139) in time order :

` Time 5.0 s:`
`        Send "set fcs/steer-cmd-norm 0.5"`
`        Send "set fcs/elevator-cmd-norm 0.5"`
`   Time 7.0 s:`
`        Send "set fcs/steer-cmd-norm 0.0"`
`        Send "set fcs/elevator-cmd-norm 0.0"`
`    Time 15.0 s: `
`        Send "set propulsion/starter_cmd 1"`
`    Time 20.0 s:`
`        Send "set fcs/throttle-cmd-norm[0] 0.6"`
`        Send "set fcs/throttle-cmd-norm[1] 0.6"`
`    Time 50.0 s:`
`        Send ""set fcs/steer-cmd-norm 1.0"`
`    Time 55.0 s:`
`        Send "set fcs/steer-cmd-norm 0.0"`
Commands above is extract from [B737_Runway.xml](jsbsim/scripts/B737_Runway.xml). But when these commands are sent to JSBSim.exe, the _vc-kts_ always stays ZERO and _position/h-agl-ft_ stays 49.0 inches all the simulation process. it isn't meet my need, some log file is below:

`----------------------JSBSim Execution Beginning----------------------`
`Start: Monday November 21 2022 15:25:56 (HH:MM:SS)`
`simulation elapsed time: 0.0083	vc-kts: 0.1589	Distance: 0.0000	ASL: 1.2446`
`simulation elapsed time: 5.0000	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`Receive() returns: set fcs/steer-cmd-norm 0.5`
`[set]	[fcs/steer-cmd-norm]	[0.5]`
`ready to reply successful.`
`Receive() returns: set fcs/elevator-cmd-norm 0.5`
`[set]	[fcs/elevator-cmd-norm]	[0.5]`
`ready to reply successful.`
`Receive() returns: set fcs/steer-cmd-norm 0.0`
`[set]	[fcs/steer-cmd-norm]	[0.0]`
`ready to reply successful.`
`Receive() returns: set fcs/elevator-cmd-norm 0.0`
`[set]	[fcs/elevator-cmd-norm]	[0.0]`
`ready to reply successful.`
`simulation elapsed time: 10.0000	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`simulation elapsed time: 15.0083	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`Receive() returns: set propulsion/starter_cmd 1`
`[set]	[propulsion/starter_cmd]	[1]`
`ready to reply successful.`
`simulation elapsed time: 20.0083	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`Receive() returns: set fcs/throttle-cmd-norm[0] 0.6`
`[set]	[fcs/throttle-cmd-norm[0]]	[0.6]`
`ready to reply successful.`
`Receive() returns: set fcs/throttle-cmd-norm[1] 0.6`
`[set]	[fcs/throttle-cmd-norm[1]]	[0.6]`
`ready to reply successful.`
`simulation elapsed time: 25.0083	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`simulation elapsed time: 30.0083	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`simulation elapsed time: 35.0083	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`simulation elapsed time: 40.0167	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`simulation elapsed time: 45.0083	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`simulation elapsed time: 50.0083	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`Receive() returns: set fcs/steer-cmd-norm 1.0`
`[set]	[fcs/steer-cmd-norm]	[1.0]`
`ready to reply successful.`
`simulation elapsed time: 55.0083	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`
`Receive() returns: set fcs/steer-cmd-norm 0.0`
`[set]	[fcs/steer-cmd-norm]	[0.0]`
`ready to reply successful.`
`simulation elapsed time: 60.0167	vc-kts: 0.0000	Distance: 0.0081	ASL: 1.1247`

In here:
ASL is taken from function `FDMExec->GetPropagate()->GetAltitudeASLmeters()`, 
Distance is taken from function `FDMExec->GetAuxiliary()->GetDistanceRelativePosition()`, 
vc-kts is taken from `FDMExec->GetAuxiliary()->GetVcalibratedKTS()`.

so : 
    1, how can I take the 737 off manually by commands? 
    2, how to write the correct command to take the plane off when an aircraft type be specified next time?
Sincerely looking forward to your help, thank you.

**Comment by seanmcleod** [comment]
Have you logged the engine thrust to confirm that the engines are actually running and producing thrust?

You seem to have missed the `propulsion/cutoff_cmd` command when you copied the existing script?

https://github.com/JSBSim-Team/jsbsim/blob/8fb77410bb8457d6bfe517e1df34660dc0df4373/scripts/B737_Runway.xml#L26-L32

**Reply by hesetone** [reply]
Ohhhhh,thank you, Sean, but there is another question. when i add `propulsion/cutoff_cmd` on, it really works. but it seems that the plane will not hold the altitude, it will climb continuously from 220s or 230s approximately. The process is plot as below image(paint by gnuplot). Now, I don't want use `ap/autopilot-roll-on`, `ap/altitude-hold` or anything else here. Can I hold the altitude just by sending command like `set fcs/xxxxxxxx 1`? 
![737_Runway_1121_0](https://user-images.githubusercontent.com/18364031/203047566-0cca45c0-21b0-4e27-b2ab-09477f163a23.png)

The image above is the result of 737 test:
vc-kts is taken from `velocities/vc-kts`
 h-agl-ft is taken from `position/h-agl-ft`
theta is taken from `attitude/theta-rad`

Actually, I did another same `737` test, in which I sent a command `set fcs/elevator-cmd-norm 0.125` to  JSBSim.exe between 220s and 230s. then the plane began to dive sharply, and it seems the plane is out control.

Back to the same question , I have a same test on aircraft `OV10`, the commands  I sent is below:
```
Time 1.0 s:
Send "set simulation/do_simple_trim 2"
Time 2.0 s:
Send "set fcs/right-brake-cmd-norm 0"
Send "set fcs/left-brake-cmd-norm 0"
Send "set propulsion/starter_cmd 1" 
Time 8.0 s: /// I guess `propulsion/engine[0]/n2` will increase over 15 8.0s later.
Send "set propulsion/cutoff_cmd 0"
Send "set fcs/throttle-cmd-norm[0] 0.4"
Send "set fcs/throttle-cmd-norm[1] 0.4"
Time 13.0 s:
Send "set fcs/throttle-cmd-norm[0] 1"
Send "set fcs/throttle-cmd-norm[1] 1"
```
Then I got result as below: 
![OV10_1121_0](https://user-images.githubusercontent.com/18364031/203060514-21cd3748-e77e-4258-a0ff-233c166458a4.png)
it seems that aircraft `OV10` didn't climbed at all. How can I take the plane `OV10` off and hold its altitude just by commands?
Thank you very much for your help.

**Comment by seanmcleod** [comment]
In general, think what a human pilot would do. So for the `OV10` case how about moving the elevator once you build up enough speed? It looks like the aircraft just continues to run along the ground at about 170kt for ever.

In terms of the 737 and holding altitude, think again what a human pilot or an auto-pilot has to do to hold altitude. They basically have to adjust the throttle and the elevator in order to trim the aircraft, and keep it in trim since as you fly fuel will be burnt off so the mass of the aircraft will decrease and the cg will shift, so you'll need to keep adjusting both throttle and elevator to keep it in trim.

So either you need to make use of an existing auto-pilot implementation, or write your own. Your only other option is to get JSBSim to perform a trim operation, which will be valid for a while until the fuel burn starts to make a noticeable difference.

**Reply by hesetone** [reply]
Thank you, Sean, I have a better understanding of this now.
